### PR TITLE
Merge from main-3

### PR DESCRIPTION
**Pull request by corneil**
- [X] Resolved compile issues, test and integration-tests.
- [X] JUnit 5, Hamcrest -> AssertJ
- [X] Spring Boot 3.x best practice.

**NOTE**: _Security Integration test fails since PR wasn't merged._

### CODE REVIEW

**Review by cppwfs — CHANGES_REQUESTED** [review]
@corneil Thanks for all the hard work here!   Most of the comments are nitpicks.

File: .github/workflows/build-snapshot-controller.yml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
               "workflow_dispatch": {
                 "owner": "spring-cloud",
                 "repo": "spring-cloud-deployer",
-                "ref": "main",
+                "ref": "main-3",
```
[the remark this answers is not in the document]
Comment (corneil) [reply .github/workflows/build-snapshot-controller.yml:32]: Maybe we should merge spring-cloud-deployer main-3 to main first.

File: build-containers.sh [hunk 2]
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
-./mvnw install -s .settings.xml -DskipTests -T 1C -am -pl :spring-cloud-dataflow-server,:spring-cloud-skipper-server,:spring-cloud-dataflow-composed-task-runner
-./mvnw spring-boot:build-image -s .settings.xml -DskipTests -T 1C -pl :spring-cloud-dataflow-server,:spring-cloud-skipper-server,:spring-cloud-dataflow-composed-task-runner
\ No newline at end of file
+./mvnw install -s .settings.xml -DskipTests -T 1C -am -pl :spring-cloud-dataflow-server,:spring-cloud-skipper-server,:spring-cloud-dataflow-composed-task-runner,:spring-cloud-dataflow-tasklauncher-sink-rabbit,:spring-cloud-dataflow-tasklauncher-sink-kafka
```
[the remark this answers is not in the document]
Comment (corneil) [reply build-containers.sh:2]: This is for building all containers locally.

File: spring-cloud-common-security-config/pom.xml [hunk 3]
```diff
@@ -2,16 +2,16 @@
...
 	<description>Spring Cloud Common Security Config</description>
 
 	<parent>
 		<groupId>org.springframework.cloud</groupId>
-		<artifactId>spring-cloud-dataflow-parent</artifactId>
-		<version>2.11.6-SNAPSHOT</version>
-		<relativePath>../spring-cloud-dataflow-parent</relativePath>
+		<artifactId>spring-cloud-dataflow-build</artifactId>
+		<version>3.0.0-SNAPSHOT</version>
+		<relativePath>../spring-cloud-dataflow-build</relativePath>
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-common-security-config/pom.xml:14]: This change was made on main-3

File: spring-cloud-dataflow-build/pom.xml [hunk 4]
```diff
@@ -52,12 +47,12 @@
 		<checkstyle.version>${maven-checkstyle-plugin.version}</checkstyle.version>
 		<puppycrawl-tools-checkstyle.version>8.29</puppycrawl-tools-checkstyle.version>
 		<spring-javaformat.version>0.0.9</spring-javaformat.version>
-		<maven-failsafe-plugin.version>3.2.3</maven-failsafe-plugin.version>
+		<maven-failsafe-plugin.version>3.0.0</maven-failsafe-plugin.version>
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-build/pom.xml]: We aren't actually using failsafe anywhere anymore.

File: spring-cloud-dataflow-build/pom.xml [hunk 5]
```diff
@@ -382,6 +396,7 @@
 						</execution>
 					</executions>
 					<configuration>
+						<loaderImplementation>CLASSIC</loaderImplementation>
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-build/pom.xml:381]: Came in from main-3

File: spring-cloud-common-security-config/spring-cloud-common-security-config-web/src/test/java/org/springframework/cloud/common/security/support/ExternalOauth2ResourceAuthoritiesMapperTests.java [hunk 6]
```diff
@@ -43,12 +43,12 @@ public class ExternalOauth2ResourceAuthoritiesMapperTests {
 	public MockWebServer mockBackEnd;
 
 
-	@BeforeEach
+	@Before
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-common-security-config/spring-cloud-common-security-config-web/src/test/java/org/springframework/cloud/common/security/support/ExternalOauth2ResourceAuthoritiesMapperTests.java:46]: I decided to leave common security as it was in main-3 in the hope that the security changes will be easier to apply.

File: spring-cloud-dataflow-build/pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-build/pom.xml]: Bumping to 3.5.0

File: spring-cloud-dataflow-classic-docs/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/BaseDocumentation.java [hunk 7]
```diff
@@ -136,14 +127,14 @@ public void setupMocks(RestDocumentationContextProvider restDocumentationContext
 	protected ApplicationContext context;
 
 	protected void prepareDocumentationTests(WebApplicationContext context,
-			RestDocumentationContextProvider restDocumentationContextProvider) throws Exception {
+											 RestDocumentationContextProvider restDocumentationContextProvider) throws Exception {
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-classic-docs/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/BaseDocumentation.java:130]: Done

File: spring-cloud-dataflow-classic-docs/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/TaskExecutionsDocumentation.java [hunk 8]
```diff
@@ -78,21 +77,20 @@ public void tearDown() throws Exception {
...
-								requestParameters(
+								queryParameters(
 										parameterWithName("name").description("The name of the task definition to launch"),
 										parameterWithName("properties")
 												.description("Application and Deployer properties to use while launching. (optional)"),
 										parameterWithName("arguments")
 												.description("Command line arguments to pass to the task. (optional)")),
 								responseFields(
 										fieldWithPath("executionId").description("The id of the task execution"),
-										fieldWithPath("schemaTarget").description("The schema target of the task state data"),
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-classic-docs/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/TaskExecutionsDocumentation.java:95]: Yes

File: spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker-junit5/pom.xml [hunk 9]
```diff
@@ -23,9 +23,14 @@
 			<artifactId>spring-cloud-dataflow-common-test-docker</artifactId>
 			<version>${project.version}</version>
 		</dependency>
+		<dependency>
+			<groupId>org.junit.jupiter</groupId>
+			<artifactId>junit-jupiter-api</artifactId>
+		</dependency>
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker-junit5/pom.xml:29]: No the whole module is a dependency included as test scope when used.

File: spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker/src/main/java/org/springframework/cloud/dataflow/common/test/docker/compose/execution/DefaultDockerCompose.java [hunk 10]
```diff
@@ -70,75 +71,78 @@ public DefaultDockerCompose(DockerComposeExecutable rawExecutable, DockerMachine
...
+		command.execute(Command.throwingOnError(), true, "kill", container.getContainerName());
 	}
 
 	@Override
 	public String exec(DockerComposeExecOption dockerComposeExecOption, String containerName,
 			DockerComposeExecArgument dockerComposeExecArgument) throws IOException, InterruptedException {
 		verifyDockerComposeVersionAtLeast(VERSION_1_7_0, "You need at least docker-compose 1.7 to run docker-compose exec");
 		String[] fullArgs = constructFullDockerComposeExecArguments(dockerComposeExecOption, containerName, dockerComposeExecArgument);
-		return command.execute(Command.throwingOnError(), fullArgs);
+		if (log.isDebugEnabled()) {
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker/src/main/java/org/springframework/cloud/dataflow/common/test/docker/compose/execution/DefaultDockerCompose.java:127]: The parameters passed to the log statement perform expensive operations

File: .github/workflows/build-snapshot-controller.yml [hunk 11]
```diff
@@ -59,7 +59,7 @@ jobs:
               "workflow_dispatch": {
                 "owner": "spring-cloud",
                 "repo": "spring-cloud-dataflow",
-                "ref": "main",
+                "ref": "main-3",
```
Comment (ilayaperumalg) [inline .github/workflows/build-snapshot-controller.yml:62]: same here: should this be `main`?

File: .github/workflows/ci-carvel.yml [hunk 12]
```diff
@@ -3,7 +3,7 @@ name: CI Carvel
 on:
   push:
     branches:
-      - main
+      - main-3
```
Comment (ilayaperumalg) [inline .github/workflows/ci-carvel.yml:6]: why do we need `main-3` still?

File: .github/workflows/build-snapshot-controller.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (cppwfs) [reply .github/workflows/build-snapshot-controller.yml:32]: Agreed!

File: spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker-junit5/pom.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (cppwfs) [reply spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker-junit5/pom.xml:29]: Gotcha.  Could we then remove the `<scope>test</scope> from the `spring-boot-starter-test` for consistency?

File: build-containers.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (cppwfs) [reply build-containers.sh:2]: Gotcha.   Just making sure we were consistent.

File: spring-cloud-dataflow-classic-docs/pom.xml [hunk 13]
```diff
@@ -114,6 +107,8 @@
 						<groupId>org.apache.maven.plugins</groupId>
 						<artifactId>maven-surefire-plugin</artifactId>
 						<configuration>
+<!--							<parallel>methods</parallel>-->
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-classic-docs/pom.xml:110]: Done

File: spring-cloud-dataflow-autoconfigure/src/test/java/org/springframework/cloud/dataflow/autoconfigure/local/AbstractSchedulerPerPlatformTest.java [hunk 14]
```diff
@@ -16,7 +16,8 @@
 
 package org.springframework.cloud.dataflow.autoconfigure.local;
 
-import io.pivotal.reactor.scheduler.ReactorSchedulerClient;
+import static org.mockito.Mockito.mock;
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-autoconfigure/src/test/java/org/springframework/cloud/dataflow/autoconfigure/local/AbstractSchedulerPerPlatformTest.java:19]: fixing

File: .github/workflows/ci-carvel.yml (hunk 12 above)
Comment (corneil) [reply .github/workflows/ci-carvel.yml:6]: updating

**Review by cppwfs — APPROVED** [review]

File: spring-cloud-common-security-config/spring-cloud-common-security-config-web/src/main/java/org/springframework/cloud/common/security/OAuthSecurityConfiguration.java [hunk 15]
```diff
@@ -104,8 +105,9 @@
  * @author Corneil du Plessis
  */
 @Configuration(proxyBeanMethods = false)
-@ConditionalOnClass(WebSecurityConfigurerAdapter.class)
-@ConditionalOnMissingBean(WebSecurityConfigurerAdapter.class)
+// SCDF 3.0 Migration - Need to re add this later with a different class or bean.
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-common-security-config/spring-cloud-common-security-config-web/src/main/java/org/springframework/cloud/common/security/OAuthSecurityConfiguration.java]: done

File: spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker-junit5/pom.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker-junit5/pom.xml:29]: fixed

File: spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker/src/test/java/org/springframework/cloud/dataflow/common/test/docker/compose/connection/PortsTests.java [hunk 16]
```diff
@@ -80,18 +79,14 @@ public void parse_actual_docker_compose_output() {
 				+ "";
 		Ports ports = Ports.parseFromDockerComposePs(psOutput, LOCALHOST_IP);
 		Ports expected = new Ports(Arrays.asList(new DockerPort(LOCALHOST_IP, 8880, 8880)));
-		assertThat(ports, is(expected));
+		assertThat(ports).isEqualTo(expected);
 	}
 
 	@Test
 	public void throw_illegal_state_exception_when_no_running_container_found_for_service() {
-		// exception.expect(IllegalArgumentException.class);
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker/src/test/java/org/springframework/cloud/dataflow/common/test/docker/compose/connection/PortsTests.java:88]: done

File: spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker/src/test/java/org/springframework/cloud/dataflow/common/test/docker/compose/execution/DockerComposeTests.java [hunk 17]
```diff
@@ -18,25 +18,23 @@
...
 import org.springframework.cloud.dataflow.common.test.docker.compose.connection.ContainerName;
 import org.springframework.cloud.dataflow.common.test.docker.compose.connection.DockerMachine;
 import org.springframework.cloud.dataflow.common.test.docker.compose.connection.DockerPort;
 import org.springframework.cloud.dataflow.common.test.docker.compose.connection.Ports;
 
-import static org.apache.commons.io.IOUtils.toInputStream;
-import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.Matchers.contains;
-import static org.hamcrest.core.Is.is;
+import static org.assertj.core.api.Assertions.*;
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-common/spring-cloud-dataflow-common-test-docker/src/test/java/org/springframework/cloud/dataflow/common/test/docker/compose/execution/DockerComposeTests.java]: fixed

**Pull request merged**